libvast/native-plugins/rebuild.cpp

```diff
@@ -234,34 +250,63 @@ rebuilder(rebuilder_actor::stateful_pointer<rebuilder_state> self,
         ->request(self->state.index, caf::infinite, atom::rebuild_v,
                   std::move(current_run_partitions))
         .then(
-          [self, rp, num_partitions, is_heterogeneous,
+          [self, rp, current_run_events, num_partitions, is_heterogeneous,
            is_oversized](std::vector<partition_info>& result) mutable {
-            // Adjust the counters, update the indicator, and move back
-            // undersized transformed partitions to the list of remainig
-            // partitions as desired.
-            VAST_ASSERT(!result.empty());
+            // Determines whether we moved partitions back.
             bool needs_second_stage = false;
-            if (is_heterogeneous) {
-              self->state.num_heterogeneous -= 1;
-              if (result.size() > 1
-                  || result[0].events <= self->state.max_partition_size / 2) {
-                std::copy(result.begin(), result.end(),
-                          std::back_inserter(self->state.remaining_partitions));
-                needs_second_stage = true;
-              }
-            } else {
+            // If the number of events in the resulting partitions does not
+            // match the number of events in the partitions that went in we ran
+            // into a conflict with other partition transformations on an
+            // overlapping set.
+            const auto detected_conflict
+              = current_run_events
+                != std::transform_reduce(result.begin(), result.end(), size_t{},
+                                         std::plus<>{},
+                                         [](const partition_info& partition) {
+                                           return partition.events;
+                                         });
```

Comment:
To be honest I don't really get the point of this. My understanding says we could just iterate over the result partitions and put them into the second stage if they are under the size threshold. Then we can remove `is_oversized`. Then again, I don't understand this algorithm as well as @dominiklohmann, and this version works. So I'm approving this.